launcher/helper.cpp

```diff
@@ -145,4 +173,157 @@ void keepScreenOn(bool isEnabled)
 	iOS_utils::keepScreenOn(isEnabled);
 #endif
 }
+
+#ifdef VCMI_ANDROID
+static constexpr int  kFolderPickerReqCode = 4242;
+
+// Intent flags: READ | WRITE | PERSISTABLE | PREFIX
+static constexpr jint kIntentFlags        = 1 | 2 | 64 | 128;
+
+class FolderPickReceiver final : public QAndroidActivityResultReceiver
+{
+public:
+    std::function<void(QString)> onDone;
+
+    // One-shot result handler for ACTION_OPEN_DOCUMENT_TREE
+    void handleActivityResult(int req, int res, const QAndroidJniObject &data) override
+    {
+        auto cb = std::move(onDone); // guarantee single-use
+
+        if (req != kFolderPickerReqCode || res != -1 /*RESULT_OK*/ || !data.isValid())
+        {
+            QMetaObject::invokeMethod(qApp, [cb]{ if (cb) cb({}); }, Qt::QueuedConnection);
+            return;
+        }
+
+        // Always return content:// tree URI
+        const QAndroidJniObject uri = data.callObjectMethod("getData","()Landroid/net/Uri;");
+        const QAndroidJniObject us  = uri.callObjectMethod("toString","()Ljava/lang/String;");
+        const QString pickedTree    = us.toString();
+
+        // Persist read+write permission
+        const QAndroidJniObject ctx = QtAndroid::androidContext();
+        const QAndroidJniObject cr  = ctx.callObjectMethod("getContentResolver","()Landroid/content/ContentResolver;");
+        cr.callMethod<void>("takePersistableUriPermission", "(Landroid/net/Uri;I)V", uri.object<jobject>(), jint(1 | 2));
+
+        // Bounce back to Qt thread
+        QMetaObject::invokeMethod(qApp, [cb, pickedTree]{ if (cb) cb(pickedTree); }, Qt::QueuedConnection);
+    }
+};
+
+static FolderPickReceiver g_receiver;
+#endif // VCMI_ANDROID
+
+void nativeFolderPicker(QWidget *parent, std::function<void(QString)> cb)
+{
+#if defined(VCMI_ANDROID)
+    Q_UNUSED(parent);
+    g_receiver.onDone = std::move(cb);
+
+    QAndroidJniObject intent("android/content/Intent","()V");
+    intent.callObjectMethod("setAction", "(Ljava/lang/String;)Landroid/content/Intent;", QAndroidJniObject::fromString("android.intent.action.OPEN_DOCUMENT_TREE").object<jstring>());
+    intent.callObjectMethod("addFlags","(I)Landroid/content/Intent;", kIntentFlags);
+
+    QtAndroid::startActivity(intent, kFolderPickerReqCode, &g_receiver);
+
+#elif defined(VCMI_IOS)
+    SelectDirectory iosDirectorySelector;
+    const QString dir = iosDirectorySelector.getExistingDirectory();
+    if (cb) cb(dir);
+
+#else
+    const QString dir = QFileDialog::getExistingDirectory(parent, {}, {}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+    if (cb) cb(dir);
```

Comment:
if `cb` is "false", then it makes no sense to select any directory at all, otherwise you select a directory for nothing